---
title: "diversity stats"
author: "Lea Stauber - Stauber et al. 2022, Mol. Ecol."
date: "1/13/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

packages <- c("BiocManager", "renv")
if (!requireNamespace(packages, quietly = TRUE))
    install.packages(packages)
rm(packages)

## to restore R environment from renv.lock file run:
# renv::restore()

library(ggplot2)
library(ggsci)
library(vcfR)
library(tidyr)
library(gridExtra)
library(dplyr)
```



### count number of SNPs and singletons
```{bash engine.opts='-l'}
## vcf files
ticino1="../data/Ticino_sampling1.genotyped.SNP.filter.PASS.mac1.biallelic.DP.GQ.AD.removeMAT.recode.vcf"
ticino2="../data/Ticino_sampling2.genotyped.SNP.filter.PASS.mac1.biallelic.DP.GQ.AD.removeMAT.recode.vcf"
noticino="../data/noTicino.genotyped.SNP.filter.PASS.maxmiss1.biallelic.mac1.rmMAT.DP.GQ.AD.final.recode.vcf"

mkdir output

## outfile names
ticino1out="output/Ticino_sampling1"
ticino2out="output/Ticino_sampling2"
noticinoout="output/noticino"

## get singleton positions
vcftools --vcf $ticino1 --singletons --out $ticino1out
vcftools --vcf $ticino2 --singletons --out $ticino2out
vcftools --vcf $noticino --singletons --out $noticinoout

```


```{r SNP counts and singletons, warning=FALSE, message=FALSE}
## get number of samples SNPs per pop
T1_SNPs <- read.vcfR("../data/Ticino_sampling1.genotyped.SNP.filter.PASS.mac1.biallelic.DP.GQ.AD.removeMAT.recode.vcf")
n_T1 <- ncol(T1_SNPs@gt)-1
T1_SNPs <- nrow(T1_SNPs@fix)
T2_SNPs <- read.vcfR("../data/Ticino_sampling2.genotyped.SNP.filter.PASS.mac1.biallelic.DP.GQ.AD.removeMAT.recode.vcf")
n_T2 <- ncol(T2_SNPs@gt)-1
T2_SNPs <- nrow(T2_SNPs@fix)
noT_SNPs <- read.vcfR("../data/noTicino.genotyped.SNP.filter.PASS.maxmiss1.biallelic.mac1.rmMAT.DP.GQ.AD.final.recode.vcf")
n_noT <- ncol(noT_SNPs@gt)-1
noT_SNPs <- nrow(noT_SNPs@fix)

## get number of singletons per pop
T1_singl <- nrow(read.table("output/Ticino_sampling1.singletons", header = F))
T2_singl <- nrow(read.table("output/Ticino_sampling2.singletons", header = F))
noT_singl <- nrow(read.table("output/noticino.singletons", header = F))


## merge into df
pop <- c("US and Europe", "Ticino 1990", "Ticino 2019")
n_samples <- c(n_noT, n_T1, n_T2)
SNPs <- c(noT_SNPs, T1_SNPs, T2_SNPs)
perc_singl <- c((noT_singl/noT_SNPs*100), (T1_singl/T1_SNPs*100), (T2_singl/T2_SNPs*100))

SNPs_df <- data.frame(pop, n_samples, SNPs, perc_singl) %>% 
  mutate(perc_singl = round(perc_singl, digits = 2))
```



### calculate nucleotide diversity using vcftools haploid (https://github.com/vcftools/vcftools/tree/fff19cd8b2f5fa13446b4d4d63e99c4fd4bf2bd2) and custom python script (D. Pereira)
```{bash engine.opts='-l'}
vcf="../data/Cp_global.genotyped.SNP.filter.PASS.mac1.biallelic.DP.GQ.AD.removeMAT.recode.vcf"

## sample files representing populations (i.e. sampling1, sampling2, Europe/USA)
ticino1="../data/Ticino.sampling1.list.txt"
ticino2="../data/Ticino.sampling2.list.txt"
noticino="../data/noTicino.list.txt"

## nuc. div outfiles
ticino1out="Ticino_sampling1"
ticino2out="Ticino_sampling2"
noticinoout="noticino"


/Users/stauber/software/vcftools-fff19cd8b2f5fa13446b4d4d63e99c4fd4bf2bd2/src/cpp/vcftools --vcf $vcf --keep $ticino1 --haploid --site-pi --out $ticino1out
/Users/stauber/software/vcftools-fff19cd8b2f5fa13446b4d4d63e99c4fd4bf2bd2/src/cpp/vcftools --vcf $vcf --keep $ticino2 --haploid --site-pi --out $ticino2out
/Users/stauber/software/vcftools-fff19cd8b2f5fa13446b4d4d63e99c4fd4bf2bd2/src/cpp/vcftools --vcf $vcf --keep $noticino --haploid --site-pi --out $noticinoout

python3 function_nucl_diversity.py .
rm *.pi
```


```{r nuc. div, message=FALSE, warning=FALSE}
nucdiv <- read.table("Final_nuc_diversity.txt", header = T) %>% 
  mutate(PopID = ifelse(grepl("noticino", PopID), "US and Europe",
                      ifelse(grepl("sampling1", PopID), "Ticino 1990", "Ticino 2019"))) %>% 
  select(c(PopID, final_genomewide_pi)) %>% 
  rename(pop = PopID)
```


### calculate SNP density and plot
```{bash engine.opts='-l'}
## vcf files
ticino1="../data/Ticino_sampling1.genotyped.SNP.filter.PASS.mac1.biallelic.DP.GQ.AD.removeMAT.recode.vcf"
ticino2="../data/Ticino_sampling2.genotyped.SNP.filter.PASS.mac1.biallelic.DP.GQ.AD.removeMAT.recode.vcf"
noticino="../data/noTicino.genotyped.SNP.filter.PASS.maxmiss1.biallelic.mac1.rmMAT.DP.GQ.AD.final.recode.vcf"

##outfiles
ticino1out="../data/Ticino_sampling1"
ticino2out="../data/Ticino_sampling2"
noticinoout="../data/noticino"

/Users/stauber/software/vcftools-fff19cd8b2f5fa13446b4d4d63e99c4fd4bf2bd2/src/cpp/vcftools --haploid --SNPdensity 1000 --vcf $ticino1 --out $ticino1out
/Users/stauber/software/vcftools-fff19cd8b2f5fa13446b4d4d63e99c4fd4bf2bd2/src/cpp/vcftools --haploid --SNPdensity 1000 --vcf $ticino2 --out $ticino2out
/Users/stauber/software/vcftools-fff19cd8b2f5fa13446b4d4d63e99c4fd4bf2bd2/src/cpp/vcftools --haploid --SNPdensity 1000 --vcf $noticino --out $noticinoout


```


```{r plot SNP density, warning=FALSE, message=FALSE}
files <- list.files(pattern="*.snpden", path = "../data")
dfnames <- gsub(".snpden", "", files)

dat <- paste0("../data/", files) %>% 
  lapply(., read.table, header = T, sep="\t", fill=T) %>% 
  Map(cbind, ., pop = dfnames) %>% 
  do.call(rbind.data.frame, .) %>% 
  mutate(CHROM = ifelse(CHROM == "mitochondria", "M", gsub("scaffold_", "", CHROM)),
         CHROM = factor(CHROM, levels = c(1:12, "M"), ordered = T),
         pop = ifelse(pop == "Ticino_sampling1", "Ticino 1990",
                      ifelse(pop == "Ticino_sampling2", "Ticino 2019", "US and Europe")),
         pop = factor(pop, levels = c("US and Europe", "Ticino 1990", "Ticino 2019"))) %>% 
  filter(VARIANTS.KB!= 0)

ggplot(dat, aes(x=BIN_START/1000, y=VARIANTS.KB, col=pop)) +
  geom_point(shape=21, alpha=.6) +
  theme_minimal() +
  facet_grid(CHROM~.) +
  scale_color_uchicago() +
  theme(legend.title = element_blank(),
        axis.text.y = element_text(size=6),
        strip.background=element_rect(fill="grey80", colour = NA)) +
  labs(x="Position [kb]", y = "SNP density") +
  scale_y_continuous(breaks = seq(0, 80, 10))

dir.create("plots")
ggsave("plots/SupplFig2_SNPdensity.pdf", useDingbats=F)

densities <- dat %>% 
  group_by(pop) %>% 
  summarise(density_range_kb = paste("0", max(VARIANTS.KB), sep = "-"))
```


### Merge diversity statistic into a single data frame
```{r summary diversity stats table, warning=FALSE, message=FALSE}
summary_df <- merge(SNPs_df, nucdiv, by="pop") 

write.table(summary_df, "output/diversity_statistics.txt", row.names = F, col.names = T, quote = F, sep = "\t")

system("rm Final_nuc_diversity.txt output/*.singletons")
```


### Calculation of allele frequencies of Ticino pops
```{bash engine.opts='-l'}
ticino1="../data/Ticino_sampling1.genotyped.SNP.filter.PASS.mac1.biallelic.DP.GQ.AD.removeMAT.recode.vcf"
ticino2="../data/Ticino_sampling2.genotyped.SNP.filter.PASS.mac1.biallelic.DP.GQ.AD.removeMAT.recode.vcf"

ticino1out="output/Ticino_sampling1"
ticino2out="output/Ticino_sampling2"


vcftools --vcf $ticino1 --freq2 --out $ticino1out
vcftools --vcf $ticino2 --freq2 --out $ticino2out

```

```{r plot ALT allele freqs, warning=FALSE, message=FALSE}
rm(list = ls())

files <- list.files(pattern="*.frq", path = "output")
dfnames <- gsub(".frq", "", files)

freq <- paste0("output/", files) %>% 
  lapply(., read.table, header = F, sep="\t", fill=T) %>% 
  lapply(., function(x) {x[-1,]}) %>% 
  Map(cbind, ., pop = dfnames) %>% 
  do.call(rbind.data.frame, .) %>% 
  select(-c(V3, V4)) %>% 
  rename(CHROM=V1, POS=V2, REFfreq=V5, ALTfreq=V6) %>% 
  mutate(pop = ifelse(pop == "Ticino_sampling1", "Sampling 1 (1990s)", "Sampling 2 (2019)"))
  
ggplot(freq, aes(x=ALTfreq, fill=pop)) +
  geom_histogram(alpha=0.7, position="identity") +
  theme_classic() +
  theme(legend.title = element_blank()) +
  scale_fill_manual(values = c("#4755A2", "#CF3C31")) +
  labs(x="Alternative allele frequency", y="Count")

ggsave("plots/Fig3A_allelefreq.pdf", height = 5, width = 5)

```


### Minor allele frequency spectra of high, moderate and modifier impact mutations 

```{r SNP impact, warning=FALSE, message=FALSE}
files <- list.files(pattern="*ANN.vcf", path = "../data")
dfnames <- gsub(".genotyped.*", "", files)

## get readable SnpEff df
snpeff <- paste0("../data/", files) %>% 
  lapply(., read.table, header = T, sep="\t", fill=T) %>% 
  Map(cbind, ., pop = dfnames) %>% 
  lapply(., function(x) {separate(x, INFO, c("V1", "V2", "V3", "V4", "V5", "V6", "V7", "V8", "V9", "V10"), "\\|")}) %>% 
  lapply(., function(x) {select(x, c(CHROM, POS, REF, ALT, V3, V4, pop))}) %>% 
  do.call(rbind.data.frame, .) %>% 
  rename(impact = V3, protein=V4) %>% 
  mutate(pop = ifelse(pop == "Ticino_sampling1", "Sampling 1 (1990s)", "Sampling 2 (2019)"))

## merge SNPeff and allele freq info and add allele frequency range for plotting
alldf <- merge(snpeff, freq, by = c("CHROM", "POS", "pop"), all = T) %>% 
  mutate(MAF = ifelse(ALTfreq >  0.5, 1 - ALTfreq, NA), # calculate minor allele freq
         CHROM = gsub("scaffold_", "", CHROM),
         CHROM = gsub("mitochondria", "M", CHROM),
         MAFrange = ifelse(MAF >= 0 & MAF <= 0.1, "0 - 0.1",
                           ifelse(MAF > 0.1 & MAF <= 0.2, "0.1 - 0.2",
                                  ifelse(MAF > 0.2 & MAF <= 0.3, "0.2 - 0.3",
                                         ifelse(MAF > 0.4 & MAF <= 0.5, "0.4 - 0.5",
                                                 NA))))) %>% 
  select(!protein) %>% 
  na.omit()


p1 <- ggplot(filter(alldf, pop == "Sampling 1 (1990s)" & impact != "LOW"), aes(x=MAFrange, fill=impact)) + 
  geom_bar(position=position_dodge2(width = 3, preserve = "single")) +
  theme_classic() +
  xlab("Minor allele frequency") +
  ggtitle("Sampling 1\n(1990s)") +
  theme(plot.title = element_text(hjust = 0.5, vjust=-1, size = 10),
        legend.title = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size=8),
        legend.key.height= unit(3, 'mm'),
        legend.key.width= unit(3, 'mm'),
        legend.text = element_text(size=8)) +
  scale_fill_futurama() 

p2 <- ggplot(filter(alldf, pop == "Sampling 2 (2019)" & impact != "LOW"), aes(x=MAFrange, fill=impact)) + 
  geom_bar(position=position_dodge2(width = 3, preserve = "single")) +
  theme_classic() +
  xlab("Minor allele frequency") +
  ggtitle("Sampling 2\n(2019)") +
  theme(plot.title = element_text(hjust = 0.5, vjust=-1, size=10),
        legend.title = element_blank(),
        axis.title.x = element_blank(),
        axis.title.y = element_text(size=8),
        legend.key.height= unit(3, 'mm'),
        legend.key.width= unit(3, 'mm'),
        legend.text = element_text(size=8)) +
  scale_fill_futurama() 


pdf("plots/Fig3B_MAF.pdf", height = 6, width = 5)
grid.arrange(p1, p2, nrow=2)
dev.off()

system("rm output/*.frq")
```

